ziti/src/main/kotlin/io/netfoundry/ziti/identity/util.kt

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2018-2020 NetFoundry, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.netfoundry.ziti.identity
+
+import io.netfoundry.ziti.util.readCerts
+import io.netfoundry.ziti.util.readKey
+import org.bouncycastle.asn1.x500.X500Name
+import org.bouncycastle.asn1.x500.style.BCStyle
+import java.io.File
+import java.net.URI
+import java.security.KeyStore
+
+/**
+ *
+ */
+internal fun findIdentityAlias(ks: KeyStore): String {
+    for (a in ks.aliases()) {
+        if (ks.isKeyEntry(a)) {
+            return a
+        }
+    }
+
+    error("no suitable key entry")
+}
+
+internal fun keystoreFromConfig(id: IdentityConfig): KeyStore {
+    val ks = KeyStore.getInstance("PKCS12")
+    ks.load(null)
+
+    val certs = readCerts(id.id.cert.replace("pem:", ""))
+    val ztAPI = URI.create(id.ztAPI)
+    val name = X500Name(certs[0].subjectDN.name).getRDNs(BCStyle.CN)[0].first.value.toASN1Primitive().toString()
+    val alias = "ziti://${ztAPI.host}:${ztAPI.port}/${name}"
+
+    val key = readKey(id.id.key.replace("pem:", "").reader())
+    val keyEntry = KeyStore.PrivateKeyEntry(key, certs.toTypedArray())
+    ks.setEntry(alias, keyEntry, KeyStore.PasswordProtection(charArrayOf()))
+
+    id.id.ca?.let {
+        val cacerts = readCerts(it.replace("pem:", ""))
+        for (ca in cacerts) {
+            val caAlias = "${alias}-ca-${ca.serialNumber}"
+            ks.setCertificateEntry(caAlias, ca)
+        }
+    }
+
+    return ks
+
+}
+
+internal fun loadKeystore(f: File, pwd: CharArray): KeyStore {
+    val ks = KeyStore.getInstance("PKCS12")
+    try {
+        ks.load(f.inputStream(), pwd)
+        return ks
+    } catch (ex: Exception) {
+        //w("failed to load $f as PKCS12 key store: $ex")
```

Comment:
commented out on purpose?

Reply:
this code was just relocated, also I don't think we need the warning there